File: React App/src/Task-6.js
// src/Task-6.js
import React, { useEffect, useState } from 'react';
import Plot from 'react-plotly.js';

const Task6 = () => {
    const [theta, setTheta] = useState(60); // Angle in degrees
    const [initialSpeed, setInitialSpeed] = useState(10); // Initial speed in m/s
    const [initialHeight, setInitialHeight] = useState(2); // Initial height in meters
    const [gravity, setGravity] = useState(9.81); // Gravity in m/s²
    const [data, setData] = useState([]);
    const [output, setOutput] = useState({});

    const zFunc = (z) => {
        return 0.5 * Math.log(Math.abs(Math.sqrt(1 + z ** 2) + z)) + 0.5 * z * Math.sqrt(1 + z ** 2);
    };

    const pcalc = (theta, u, g, h, N) => {
        const thetaRad = theta * (Math.PI / 180); // Convert to radians
        const R = ((u ** 2) / g) * (Math.sin(thetaRad) * Math.cos(thetaRad) + Math.cos(thetaRad) * Math.sqrt(Math.sin(thetaRad) ** 2 + 2 * g * h / (u ** 2)));
        const x = Array.from({ length: N }, (_, i) => (i / (N - 1)) * R);
        const t = x.map(xi => xi / (u * Math.cos(thetaRad)));
        const T = R / (u * Math.cos(thetaRad));
        const y = x.map(xi => h + xi * Math.tan(thetaRad) - (g / (2 * u ** 2)) * (xi ** 2) * (1 + Math.tan(thetaRad) ** 2));
        const ta = u * Math.sin(thetaRad) / g;
        const xa = (u ** 2) * Math.sin(2 * thetaRad) / (2 * g);
        const ya = h + ((u ** 2) / (2 * g)) * Math.sin(thetaRad) ** 2;
        const vx = Array(N).fill(u * Math.cos(thetaRad));
        const vy = vx.map((_, i) => u * Math.sin(thetaRad) - g * t[i]);
        const v = vx.map((_, i) => Math.sqrt(vx[i] ** 2 + vy[i] ** 2));
        const phi = vx.map((_, i) => Math.atan2(vy[i], vx[i]));

        const a = (u ** 2) / (g * (1 + (Math.tan(thetaRad)) ** 2));
        const b = Math.tan(thetaRad);
        const c = Math.tan(thetaRad) - g * R * (1 + (Math.tan(thetaRad)) ** 2) / (u ** 2);
        const s = a * (zFunc(b) - zFunc(c));

        const dx = x.slice(1).map((xi, i) => xi - x[i]);
        const dy = y.slice(1).map((yi, i) => yi - y[i]);
        const sNumeric = dx.reduce((sum, dx, i) => sum + Math.sqrt(dx ** 2 + dy[i] ** 2), 0);
        const thetaM = Math.asin(Math.sqrt(1 / (2 + 2 * g * h / (u ** 2)))) * (180 / Math.PI);
        const Tm = (u / g) * Math.sqrt(2 + 2 * g * h / (u ** 2));
        const Rm = ((u ** 2) / g) * Math.sqrt(1 + 2 * g * h / (u ** 2));

        return { R, x, y, T, ta, xa, ya, s, sNumeric, thetaM, Tm, Rm };
    };

    useEffect(() => {
        const pGiven = pcalc(theta, initialSpeed, gravity, initialHeight, 500);
        const thetaOptimal = Math.asin(1 / Math.sqrt(2 + 2 * gravity * initialHeight / (initialSpeed ** 2))) * (180 / Math.PI);
        const pOptimal = pcalc(thetaOptimal, initialSpeed, gravity, initialHeight, 500);

        setData([
            {
                x: pGiven.x,
                y: pGiven.y,
                type: 'scatter',
                mode: 'lines',
                line: { color: 'blue' },
                name: `Given Angle: ${theta} degrees`,
            },
            {
                x: pOptimal.x,
                y: pOptimal.y,
                type: 'scatter',
                mode: 'lines',
                line: { color: 'red', dash: 'dash' },
                name: `Optimal Angle: ${thetaOptimal.toFixed(2)} degrees`,
            }
        ]);

        setOutput({
            'Given Trajectory': {
                'Range': `${pGiven.R.toFixed(2)} m`,
                'Time of flight': `${pGiven.T.toFixed(2)} s`,
                'Apogee': `(${pGiven.xa.toFixed(2)}, ${pGiven.ya.toFixed(2)}) m`,
                'Length of trajectory (analytical)': `${pGiven.s.toFixed(2)} m`,
                'Length of trajectory (numeric)': `${pGiven.sNumeric.toFixed(2)} m`,
            },
            'Optimal Trajectory for Max Range': {
                'Range': `${pOptimal.R.toFixed(2)} m`,
                'Time of flight': `${pOptimal.T.toFixed(2)} s`,
                'Apogee': `(${pOptimal.xa.toFixed(2)}, ${pOptimal.ya.toFixed(2)}) m`,
                'Length of trajectory (analytical)': `${pOptimal.s.toFixed(2)} m`,
                'Length of trajectory (numeric)': `${pOptimal.sNumeric.toFixed(2)} m`,
            }
        });
    }, [theta, initialSpeed, initialHeight, gravity]);

    return (
        <div style={{ display: 'flex', flexDirection: 'column', alignItems: 'flex-start' }}>
            <h1>Task 6 - Projectile Trajectories</h1>
            <div style={{ display: 'flex' }}>
                <div style={{ flex: 1 }}>
                    <Plot
                        data={data}
                        layout={{
                            title: 'Projectile Trajectories',
                            xaxis: { title: 'Horizontal Distance (m)' },
                            yaxis: { title: 'Vertical Distance (m)' },
                        }}
                        style={{ width: '100%', height: '500px' }}
                    />
                </div>
                <div style={{ marginLeft: '20px' }}>
                    {Object.entries(output).map(([key, values]) => (
                        <div key={key} style={{ marginBottom: '20px' }}>
                            <h2>{key}</h2>
                            <ul>
                                {Object.entries(values).map(([subKey, value]) => (
                                    <li key={subKey}><strong>{subKey}:</strong> {value}</li>
                                ))}
                            </ul>
                        </div>
                    ))}
                </div>
            </div>
            <div>
                <label>
                    Launch Angle (degrees):
                    <input
                        type="range"
                        min="0"
                        max="90"
                        step="0.01"
                        value={theta}
                        onChange={(e) => setTheta(parseFloat(e.target.value))}
                    />
                    <input
                        type="number"
                        min="0"
                        max="90"
                        step="0.01"
                        value={theta.toFixed(2)}
                        onChange={(e) => setTheta(parseFloat(e.target.value))}
                        style={{ width: '100px', marginLeft: '10px' }}
                    />
                </label>
            </div>
            <div>
                <label>
                    Initial Speed (m/s):
                    <input
                        type="range"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={initialSpeed}
                        onChange={(e) => setInitialSpeed(parseFloat(e.target.value))}
                    />
                    <input
                        type="number"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={initialSpeed.toFixed(2)}
                        onChange={(e) => setInitialSpeed(parseFloat(e.target.value))}
                        style={{ width: '100px', marginLeft: '10px' }}
                    />
                </label>
            </div>
            <div>
                <label>
                    Initial Height (m):
                    <input
                        type="range"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={initialHeight}
                        onChange={(e) => setInitialHeight(parseFloat(e.target.value))}
                    />
                    <input
                        type="number"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={initialHeight.toFixed(2)}
                        onChange={(e) => setInitialHeight(parseFloat(e.target.value))}
                        style={{ width: '100px', marginLeft: '10px' }}
                    />
                </label>
            </div>
            <div>
                <label>
                    Gravity (m/s²):
                    <input
                        type="range"
                        min="1"
                        max="20"
                        step="0.01"
                        value={gravity}
                        onChange={(e) => setGravity(parseFloat(e.target.value))}
                    />
                    <input
                        type="number"
                        min="1"
                        max="20"
                        step="0.01"
                        value={gravity.toFixed(2)}
                        onChange={(e) => setGravity(parseFloat(e.target.value))}
                        style={{ width: '100px', marginLeft: '10px' }}
                    />
                </label>
            </div>
        </div>
    );
};

export default Task6;

File: Tasks in Python/Code/Task_6.py
# Required libraries
import numpy as np
import matplotlib.pyplot as plt

# Function to compute the z function used in trajectory length calculation
def z_func(z):
    return 0.5 * np.log(np.abs(np.sqrt(1 + z ** 2) + z)) + 0.5 * z * np.sqrt(1 + z ** 2)

# Function to calculate various properties of the projectile motion
def pcalc(theta, u, g, h, N):
    p = {}
    theta_rad = np.radians(theta)  # Convert angle to radians

    # Calculate range of the projectile
    p['R'] = ((u ** 2) / g) * (np.sin(theta_rad) * np.cos(theta_rad) + np.cos(theta_rad) * np.sqrt(np.sin(theta_rad) ** 2 + 2 * g * h / (u ** 2)))

    # Compute the trajectory points
    p['x'] = np.linspace(0, p['R'], N)  # Horizontal distances
    p['t'] = p['x'] / (u * np.cos(theta_rad))  # Time of flight for each horizontal distance
    p['T'] = p['R'] / (u * np.cos(theta_rad))  # Total time of flight
    p['y'] = h + p['x'] * np.tan(theta_rad) - (g / (2 * u ** 2)) * (p['x'] ** 2) * (1 + np.tan(theta_rad) ** 2)  # Vertical distance
    
    # Compute apogee and velocity components
    p['ta'] = u * np.sin(theta_rad) / g
    p['xa'] = (u ** 2) * np.sin(2 * theta_rad) / (2 * g)
    p['ya'] = h + ((u ** 2) / (2 * g)) * np.sin(theta_rad) ** 2
    p['vx'] = u * np.cos(theta_rad) * np.ones(N)  # Horizontal velocity
    p['vy'] = u * np.sin(theta_rad) - g * p['t']  # Vertical velocity
    p['v'] = np.sqrt(p['vx'] ** 2 + p['vy'] ** 2)  # Total velocity
    p['phi'] = np.arctan2(p['vy'], p['vx'])  # Angle of velocity vector

    # Compute the analytical length of the trajectory
    a = (u ** 2) / (g * (1 + (np.tan(theta_rad)) ** 2))
    b = np.tan(theta_rad)
    c = np.tan(theta_rad) - g * p['R'] * (1 + (np.tan(theta_rad)) ** 2) / (u ** 2)
    p['s'] = a * (z_func(b) - z_func(c))

    # Compute the numerical length of the trajectory using numerical integration
    dx = np.diff(p['x'])
    dy = np.diff(p['y'])
    p['s_numeric'] = np.sum(np.sqrt(dx ** 2 + dy ** 2))

    # Compute properties for the maximum range trajectory
    p['theta_m'] = np.degrees(np.arcsin(np.sqrt(1 / (2 + 2 * g * h / (u ** 2)))))  # Optimal launch angle for maximum range
    p['T_m'] = (u / g) * np.sqrt(2 + 2 * g * h / (u ** 2))  # Total time of flight for maximum range
    p['R_m'] = ((u ** 2) / g) * np.sqrt(1 + 2 * g * h / (u ** 2))  # Maximum range

    return p

# Parameters for the projectile
theta = 60  # Given launch angle (degrees)
u = 10      # Initial launch speed (m/s)
g = 9.81    # Acceleration due to gravity (m/s^2)
h = 2       # Initial height (m)
N = 500     # Number of points in trajectory calculation

# Calculate properties for the given and optimal trajectories
p_given = pcalc(theta, u, g, h, N)
theta_optimal = np.degrees(np.arcsin(np.sqrt(1 / (2 + 2 * g * h / (u ** 2)))))  # Optimal angle for maximum range
p_optimal = pcalc(theta_optimal, u, g, h, N)

# Plotting the trajectories
plt.figure(figsize=(12, 6))
plt.plot(p_given['x'], p_given['y'], label=f'Trajectory: θ={theta}°')
plt.plot(p_optimal['x'], p_optimal['y'], label=f'Max Range Trajectory: θ={theta_optimal:.2f}°', linestyle='--')
plt.xlabel('Horizontal Distance (m)')
plt.ylabel('Vertical Distance (m)')
plt.title('Projectile Trajectories')
plt.legend()
plt.grid(True)
plt.show()

# Output the calculated properties
print(f"Given Trajectory:")
print(f"  Range: {p_given['R']:.2f} m")
print(f"  Time of flight: {p_given['T']:.2f} s")
print(f"  Apogee: ({p_given['xa']:.2f}, {p_given['ya']:.2f}) m")
print(f"  Length of trajectory (analytical): {p_given['s']:.2f} m")
print(f"  Length of trajectory (numeric): {p_given['s_numeric']:.2f} m")

print(f"Optimal Trajectory for Max Range:")
print(f"  Range: {p_optimal['R']:.2f} m")
print(f"  Time of flight: {p_optimal['T']:.2f} s")
print(f"  Apogee: ({p_optimal['xa']:.2f}, {p_optimal['ya']:.2f}) m")
print(f"  Length of trajectory (analytical): {p_optimal['s']:.2f} m")
print(f"  Length of trajectory (numeric): {p_optimal['s_numeric']:.2f} m")
